openedx/core/djangoapps/programs/tests/test_utils.py

```diff
@@ -967,6 +980,144 @@ def test_multiple_published_course_runs(self):
+        """
+        course1 = _create_course(self, self.course_price, course_run_count=0)
+        course2 = _create_course(self, self.course_price)
+        course2['entitlements'].append(EntitlementFactory(mode=CourseMode.PROFESSIONAL))
+        program = ProgramFactory(
+            courses=[course1, course2],
+            is_program_eligible_for_one_click_purchase=True,
+            applicable_seat_types=[CourseMode.VERIFIED],
+        )
+        data = ProgramDataExtender(program, self.user).extend()
+        self.assertFalse(data['is_learner_eligible_for_one_click_purchase'])
+        self.assertEqual(data['skus'], [])
+
+    def test_entitlement_product_and_user_enrollment(self):
+        """
+        Learner should be eligible for one click purchase if they hold an enrollment
+        but not an entitlement in a course for which there exists an entitlement product.
+        """
+        course1 = _create_course(self, self.course_price, make_entitlement=True)
+        course2 = _create_course(self, self.course_price)
+        expected_skus = set([course2['course_runs'][0]['seats'][0]['sku']])
+        CourseEnrollmentFactory(user=self.user, course_id=course1['course_runs'][0]['key'], mode=CourseMode.VERIFIED)
+        program = ProgramFactory(
+            courses=[course1, course2],
+            is_program_eligible_for_one_click_purchase=True,
+            applicable_seat_types=[CourseMode.VERIFIED],
+        )
+        data = ProgramDataExtender(program, self.user).extend()
+        self.assertTrue(data['is_learner_eligible_for_one_click_purchase'])
+        self.assertEqual(set(data['skus']), expected_skus)
+
+    def test_no_id_professional_entitlement_product(self):
+        """
+        Learner should be eligible for one click purchase if 'professional' is an applicable seat type
+        a remaining unpurchased course has a 'no-id-professional' entitlement.
+        """
+        course1 = _create_course(self, self.course_price, make_entitlement=True)
+        course2 = _create_course(self, self.course_price)
+        course2['entitlements'].append(EntitlementFactory(mode=CourseMode.NO_ID_PROFESSIONAL_MODE))
+        expected_skus = set([course1['entitlements'][0]['sku'], course2['entitlements'][0]['sku']])
+        program = ProgramFactory(
+            courses=[course1, course2],
+            is_program_eligible_for_one_click_purchase=True,
+            applicable_seat_types=[CourseMode.VERIFIED, CourseMode.PROFESSIONAL],
+        )
+        data = ProgramDataExtender(program, self.user).extend()
+        self.assertTrue(data['is_learner_eligible_for_one_click_purchase'])
+        self.assertEqual(set(data['skus']), expected_skus)
+
+    def test_no_id_professional_user_entitlement(self):
+        """
+        Learner should be eligible for one click purchase if 'professional' is an applicable seat type
+        and they hold a 'no-id-professional' entitlement.
```

Comment:
As mentioned above, we shouldn't have `no-id-professional` entitlements. It might be possible for a user to have a `no-id-professional` enrollment, however, so we should verify that a user with such an enrollment would still be eligible for one-click purchase if there are other courses with valid entitlements.